lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -1172,7 +1164,10 @@ private function isLimitQuery() : bool
      */
     private function getSQLForInsert() : string
     {
-        return 'INSERT INTO ' . $this->sqlParts['from']['table'] .
+        /** @var From $from */
+        $from = $this->sqlParts['from'][0];
```

Comment:
I'll quote my PR #3829 here:

> The original code seemed to accept either a single value, or an array of values, in `$this->sqlParts['table']` (each value being an associative array with `'table'` and `'alias'` keys).
>
> But in this method, and a few others, **it seems to assume that there is always a single value**, which is confirmed by the tests passing when I always use `$this->queryParts->from[0]`, and failing if I use another index.
>
> **So it looks like this method will fail badly if there's not exactly one table registered.** This PR **does not change this assumption**, but it would be good to ensure that this assumption is always true when this method is called.

Reply:
Not sure why changing an array with an object requires this change (same as above).

Reply:
> So it looks like this method will fail badly if there's not exactly one table registered.

Most likely, it will, which is a bug. If the object is in an invalid state, it should throw an exception. Multiple `from()` tables seem to be acceptable only for `SELECT`.

Reply:
After some more research, it looks like for `select()`, the `'from'` part may take multiple `From`s in a zero-indexed array while for `insert()`, `update()` and `delete()`, it's always one table. The question is, why do we need a `From` with a table and an alias for non-`select()` queries? They don't even use the `FROM` keyword in their syntax.

We may try separating the two concerns into two different query parts (array of `From`s in `'from'`) and a single table in `'table'`.

Reply:
What about `DELETE a FROM a, b WHERE ...`?

Reply:
> What about `DELETE a FROM a, b WHERE ...`?

Is it currently supported? FWIW, it's not portable. Usually, adding new features during refactoring is not the best idea.

Reply:
> Is it currently supported? FWIW, it's not portable.

Looks like it isn't, and may be MySQL-specific indeed.

> The question is, why do we need a From with a table and an alias for non-select() queries?
> They don't even use the FROM keyword in their syntax.

`DELETE` does use `FROM`, so I guess it *could* make sense to use it, but explicitly restrict it to a single element.

That being said, if the aim is to have `DELETE`, `INSERT` and `UPDATE` all take a single table, we should indeed probably enforce this in another property (or array key, for now).

And I agree, we probably don't need/want an alias in `DELETE`, `INSERT` and `UPDATE`. First of all, it's not very useful when there's a single table. And, AFAICS, this is not portable either. I tested aliases on db-fiddle on 3 dbs:

| | MySQL | PostgreSQL | SQLite |
| ------------- |:-------------:| -----:| -----:|
| `DELETE` alias | ❌ | ✔️ | ✔️ |
| `INSERT` alias | ❌ | ✔️ | ✔️ |
| `UPDATE` alias | ✔️ | ❌ | ❌ |

So we could just ditch aliases for these, and just add a new `'table'` entry accepting the table name only, as `string|null`?

So basically:

| Query type | Key/Prop | Type
| ------------- |:-------------:| -----:|
| `SELECT` | `'from'` | `From[]` |
| `INSERT`, `UPDATE`, `DELETE` | `'table'` | `string\|null` |

Reply:
Here's how it could look: 58ae117a0e29270e88a22f5681add92c16a6b48d.

Note how the `QueryBuilder` method signatures change, though:

| Before | After |
| ------------- |:-------------:|
| `delete(?string $delete = null, ?string $alias = null)` | `delete(string $delete)` |
| `update(?string $update = null, ?string $alias = null)` | `update(string $update)` |
| `insert(?string $insert = null)` | `insert(string $insert)` |

I believe this change is necessary for 2 reasons:

- **nullable table name**: currently "supported", but throws an Error when you attempt to get the SQL of a `QueryBuilder` with an empty `insert()`, `update()`, or `delete()`.
- **alias**: as mentioned above, it brings close to zero value to have an alias for `INSERT`, `UPDATE` or `DELETE` statement, involving a single table.

Note that I would also rename the `$delete`, `$update` and `$insert` parameters to `$table` for consistency and readability.

@morozov, what do you think?

Reply:
> `DELETE` does use `FROM`, so I guess it *could* make sense to use it, but explicitly restrict it to a single element.

When would it make sense? Maybe we could give the `From` class a better name, e.g. `TableWithAlias`. This way, it will be obvious that it's only for select.

Reply:
> @morozov, what do you think?

Makes perfect sense. 

> Note that I would also rename the `$delete`, `$update` and `$insert` parameters to `$table` for consistency and readability.

Please do.

Reply:
Another reason to enforce the `$table` argument in statements:
```php
echo $conn->createQueryBuilder()
    ->delete()
    ->from('foo')

// PHP Notice:  Undefined index: alias in lib/Doctrine/DBAL/Query/QueryBuilder.php on line 1205
// PHP Notice:  Undefined index: table in lib/Doctrine/DBAL/Query/QueryBuilder.php on line 1206
// DELETE FROM
```

Reply:
> Please do.

Done in https://github.com/doctrine/dbal/pull/3833/commits/d5ad2b5cea9203cc212f22420e3787778080e8ca